src/pages/spawn/spawnVolume/spawnVolumeTable/spawnVolumeTableActions/editButton/EditVolumeModal.tsx

```diff
@@ -0,0 +1,128 @@
+import React, { useReducer } from "react";
+import { useMutation } from "@apollo/client";
+import { Variant } from "@leafygreen-ui/button";
+import { Input } from "antd";
+import { useSpawnAnalytics } from "analytics";
+import { Modal } from "components/Modal";
+import {
+  ExpirationField,
+  ModalContent,
+  SectionContainer,
+  SectionLabel,
+  WideButton,
+} from "components/Spawn";
+import { InputLabel } from "components/styles";
+import { useBannerDispatchContext } from "context/banners";
+import {
+  UpdateVolumeMutation,
+  UpdateVolumeMutationVariables,
+} from "gql/generated/types";
+import { UPDATE_SPAWN_VOLUME } from "gql/mutations";
+import { MyVolume } from "types/spawn";
+import { reducer } from "./editVolumeModal/reducer";
+
+interface Props {
+  visible: boolean;
+  onCancel: () => void;
+  volume: MyVolume;
+}
+
+export const EditVolumeModal: React.FC<Props> = ({
+  visible,
+  onCancel,
+  volume,
+}) => {
+  const [state, dispatch] = useReducer(reducer, {
+    expiration: volume.expiration,
+    volumeId: volume.id,
+    noExpiration: volume.noExpiration,
+    name: volume.displayName,
+  });
+  const dispatchBanner = useBannerDispatchContext();
+  const spawnAnalytics = useSpawnAnalytics();
+  const [updateVolumeMutation, { loading }] = useMutation<
+    UpdateVolumeMutation,
+    UpdateVolumeMutationVariables
+  >(UPDATE_SPAWN_VOLUME, {
+    onCompleted() {
+      onCancel();
+      dispatchBanner.successBanner("Successfully updated volume");
+    },
+    onError(err) {
+      onCancel();
+      dispatchBanner.errorBanner(
+        `There was an error while updating your volume: ${err.message}`
+      );
+    },
+    refetchQueries: ["myVolumes"],
+  });
+
+  const updateVolume = () => {
+    const mutationVars = { ...state };
+    if (mutationVars.name === volume.displayName) {
+      delete mutationVars.name;
+    }
+    if (mutationVars.noExpiration === true) {
+      delete mutationVars.expiration;
+    }
+    const variables = { UpdateVolumeInput: mutationVars };
+    spawnAnalytics.sendEvent({
+      name: "Edited a Spawn Volume",
+      params: variables,
+    });
+    updateVolumeMutation({
+      variables,
+    });
+  };
+
+  const noChange =
```

Comment:
Does it make sense to make a custom hook that checks if the state is dirty? Maybe one that takes in an initial state and a current state, and does a deep compare of them? I imagine the pattern we're doing here could be generalized and used in a couple places we have forms

Reply:
we actually have access to an [isEqual function](https://lodash.com/docs/4.17.15#isEqual) that does a deep compare for objects. Using the function could have cleaned up some code here but the reason we couldn't use it was because the mutation variables require a Date obj for `expiration` where as the the `expiration` field in the Volume gql type actually returns as a string from MyVolumes query.